### PR TITLE
Fix sceneview draw modes for baked GI

### PR DESCRIPTION
**Pull request by nerites**
# **Please read the [Contributing guide](CONTRIBUTING.md) before making a PR.**

* Read the [Graphics repository & Yamato FAQ](http://go/graphics-yamato-faq).

### Checklist for PR maker
- [x] Have you added a backport label (if needed)? For example, the `need-backport-*` label. After you backport the PR, the label changes to `backported-*`.
- [x] Have you updated the changelog? Each package has a `CHANGELOG.md` file.
- [ ] Have you updated or added the documentation for your PR? When you add a new feature, change a property name, or change the behavior of a feature, it's best practice to include related documentation changes in the same PR. If you do add documentation, make sure to add the relevant Graphics Docs team member as a reviewer of the PR. If you are not sure which person to add, see the [Docs team contacts sheet](https://docs.google.com/spreadsheets/d/1rgUWWgwLFEHIQ3Rz-LnK6PAKmbM49DZZ9al4hvnztOo/edit#gid=1058860420).
- [N/A] Have you added a graphic test for your PR (if needed)? When you add a new feature, or discover a bug that tests don't cover, please add a graphic test.

---
### Purpose of this PR
Why is this PR needed, what hard problem is it solving/fixing?
Fix for underlying issue behind https://fogbugz.unity3d.com/f/cases/1211436/
Although both SRPs have provided meta passes for their shaders, they weren't being picked up or used to render the sceneview draw modes. Instead, rendering was falling back to a built-in shader that did not work for instanced terrain. 

This PR adds a common MetaPass.hlsl file used for both Universal and HD sceneview draw mode rendering. Since Builtin's PBR albedo validation and PBR specular validation modes are unsupported in SRP, this file omitted the code used only for those modes. 

Realtime GI behavior is unreliable, but matches builtin behavior (at least with tip of trunk), so I'm going ahead with just the baked GI fixes for now.

I will be making additional changes to shadergraph as well, but I'll have that as a separate PR unless anyone objects.

---
### Testing status
Manually checked behavior against built-in (tip of trunk) for baked GI draw modes using progressive CPU.

Known inconsistencies:

- Terrain doesn't render in the baked lightmap culling mode for URP and built-in.  -> Fixed in URP by e7928d7
- Emissives don't render properly for HDRP. However this may be an issue with my scene setup, since emission doesn't appear to have an effect during normal rendering, either.
- The shadowmask debug mode is disabled for HDRP in the menu. Can someone from HDRP let me know if this is intentional?
- Shadergraphs may not behave as expected with this functionality, since I'm not sure if they add the meta pass with the appropriate. Will seek feedback about how to make the proper changes to enable this for shadergraph

Current status image:
image

---
### Comments to reviewers
Notes for the reviewers you have assigned.
@sebastienlagarde and @phi-lira can you please assign this PR to someone on your team with bandwidth?

### CODE REVIEW

**Review by iM0ve — APPROVED** [review]
I have checked on HDRP the baked view is fixed.
Note that other views like Texel Validity, Albedo, Emissive will require a separate fix


Baked view before the PR in the template
Unity_STcME5fRVq


Baked view after the PR in the template
Unity_BA2MFYJp94

**Comment by sebastienlagarde** [comment]
>Emissives don't render properly for HDRP. However this may be an issue with my scene setup, since emission doesn't appear to have an effect during normal rendering, either.

Maybe your emissive intensity is 0? We have graphics test for all kind of emissive both bake and no bake, so we know it works. I expect you have an intensity of zero, this can be checked in our render pipelin debug windows with lighting mode -> Emissive lighting only.

>The shadowmask debug mode is disabled for HDRP in the menu. Can someone from HDRP let me know if this is intentional?

Shadowmask is disabled by default so I suppose you just forget to enable it?

File: com.unity.render-pipelines.high-definition/Runtime/Material/AxF/ShaderPass/AxFSharePass.hlsl
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/Material/AxF/ShaderPass/AxFSharePass.hlsl:20]: I don't see any UV3 in the meta pass, are you sure you need this?  note that the meta pass used uv0, uv1 and uv2 but no uv3

File: com.unity.render-pipelines.high-definition/Runtime/Material/Unlit/Unlit.hlsl
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/Material/Unlit/Unlit.hlsl:59]: I suppose it is to match builtin behavior?   We chose to not handle color in HDRP as color don't have any light value. HDRP rely on physical light value and thus, only Emissive color have it, we should avoid to have regular color in it. (Also I expect it may break various project)

**Review by sebastienlagarde — COMMENTED** [review]
Thanks to look at this issue and for the various change. few questoin on my side.

File: com.unity.render-pipelines.universal/ShaderLibrary/MetaInput.hlsl
Comment (phi-lira) [inline com.unity.render-pipelines.universal/ShaderLibrary/MetaInput.hlsl:84]: This is probably merge conflict? It should be removed as we removed LWRP upgrade tooling here: https://github.com/Unity-Technologies/Graphics/pull/3051/files#diff-42c6b6ac3d960a42860bdc7f7b585ebe5c4f1f2580fa2fde5e805d33e05f92e3L49-L59

File: com.unity.render-pipelines.universal/ShaderLibrary/MetaInput.hlsl
Comment (phi-lira) [inline com.unity.render-pipelines.universal/ShaderLibrary/MetaInput.hlsl:28]: There might be users using this, you could still define this and redirect to UnityMetaVertexPosition for backward compatibility.

File: com.unity.render-pipelines.universal/ShaderLibrary/MetaInput.hlsl
Comment (phi-lira) [inline com.unity.render-pipelines.universal/ShaderLibrary/MetaInput.hlsl:21]: There could be users declaring this in their custom shaders, you could forward declare this to UnityMetaInput for backward compatibility / avoid breaking changes.

File: com.unity.render-pipelines.universal/ShaderLibrary/MetaInput.hlsl
Comment (phi-lira) [inline com.unity.render-pipelines.universal/ShaderLibrary/MetaInput.hlsl:47]: Same, for backward compatibility forward declare this to `UnityMetaFragment`

File: com.unity.render-pipelines.universal/Shaders/BakedLitMetaPass.hlsl
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Shaders/BakedLitMetaPass.hlsl:7]: Also probably merge conflict? was removed here: https://github.com/Unity-Technologies/Graphics/pull/3051/files#diff-9faa832cd3ae7ccf303e2c5580cba1ccaad39c4dbaac548dfe802d33a985b4a5L41-L50

File: com.unity.render-pipelines.universal/Shaders/BakedLit.shader
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Shaders/BakedLit.shader:226]: Not sure why these are changing?

File: com.unity.render-pipelines.universal/Shaders/Lit.shader
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Shaders/Lit.shader:337]: This is only needed in editor, how is this keyword stripped?

File: com.unity.render-pipelines.universal/Shaders/LitMetaPass.hlsl
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Shaders/LitMetaPass.hlsl:23]: Should be removed

File: com.unity.render-pipelines.universal/Shaders/LitMetaPass.hlsl
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Shaders/LitMetaPass.hlsl:6]: Not sure I follow why renaming this and why needed to rename `Varyings` to `MetaVaryings`.

File: com.unity.render-pipelines.universal/ShaderLibrary/MetaInput.hlsl
Comment (phi-lira) [inline com.unity.render-pipelines.universal/ShaderLibrary/MetaInput.hlsl:28]: I'm not sure why the rename to MetaVaryings is needed and I worry that now user could be using this as thsi is part of `ShaderLibrary` and will have compiler errors. At the bare least you could forward declare the old to the new MetaVaryings if needed.

File: com.unity.render-pipelines.universal/Shaders/SimpleLitMetaPass.hlsl
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Shaders/SimpleLitMetaPass.hlsl:21]: Should be removed.

File: com.unity.render-pipelines.universal/Shaders/SimpleLitMetaPass.hlsl
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Shaders/SimpleLitMetaPass.hlsl:6]: Same here, not sure why renamed.

File: com.unity.render-pipelines.universal/Shaders/UnlitMetaPass.hlsl
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Shaders/UnlitMetaPass.hlsl:14]: Should be removed.

**Review by phi-lira — CHANGES_REQUESTED** [review]

File: com.unity.render-pipelines.high-definition/Runtime/Material/AxF/ShaderPass/AxFSharePass.hlsl
Comment (nerites) [reply com.unity.render-pipelines.high-definition/Runtime/Material/AxF/ShaderPass/AxFSharePass.hlsl:20]: The meta pass needs VARYINGS_NEED_TEXCOORD3 defined. So I guess I should move this down below as follows:


#ifdef ATTRIBUTES_NEED_TEXCOORD3  || ((SHADERPASS == SHADERPASS_LIGHT_TRANSPORT) && defined(EDITOR_VISUALIZATION))
#define VARYINGS_NEED_TEXCOORD3
#endif

File: com.unity.render-pipelines.high-definition/Runtime/Material/Unlit/Unlit.hlsl
Comment (nerites) [reply com.unity.render-pipelines.high-definition/Runtime/Material/Unlit/Unlit.hlsl:59]: Right, in builtin, unlit shaders still render out the diffuse color. If HDRP intends to have different behavior for this view, then I'm ok to revert this.

File: com.unity.render-pipelines.universal/Shaders/BakedLit.shader
Comment (nerites) [reply com.unity.render-pipelines.universal/Shaders/BakedLit.shader:226]: Since BakedLit and MetaUnlit did the same thing in their fragment shader, I had BakedLit reuse the Unlit fragment shader. Would you prefer I keep BakedLit separate?

File: com.unity.render-pipelines.universal/Shaders/Lit.shader
Comment (nerites) [reply com.unity.render-pipelines.universal/Shaders/Lit.shader:337]: How can I make sure it gets stripped from builds?

File: com.unity.render-pipelines.high-definition/Runtime/Material/AxF/ShaderPass/AxFSharePass.hlsl
Comment (nerites) [reply com.unity.render-pipelines.high-definition/Runtime/Material/AxF/ShaderPass/AxFSharePass.hlsl:20]: Ah, VertMesh requires that uv3 exists if texCoord3 is used. So I will leave it as is if that's okay.

**Comment by nerites** [comment]
> 
> 
> > Emissives don't render properly for HDRP. However this may be an issue with my scene setup, since emission doesn't appear to have an effect during normal rendering, either.
> 
> Maybe your emissive intensity is 0? We have graphics test for all kind of emissive both bake and no bake, so we know it works. I expect you have an intensity of zero, this can be checked in our render pipelin debug windows with lighting mode -> Emissive lighting only.
> 
> > The shadowmask debug mode is disabled for HDRP in the menu. Can someone from HDRP let me know if this is intentional?
> 
> Shadowmask is disabled by default so I suppose you just forget to enable it?

Ah, thank you! When I enabled shadowmask, the debug view works correctly in HDRP:
image

I'm having a bit more trouble with emissives though. It looks like different shaders handle emission differently, but I enabled emission maps wherever I could, and chose a default checkerboard texture to use as the emission map. All emission colors are non-black, and all emission weights are 1. 

When I render in HDRP's emissive debug mode, only the unlits are rendered in color (even the Unlit that has emission set to black). Everything else has no emission. 
Do HDRP's debug modes also rely on the meta pass? Is it possible I've unintentionally modified HDRP's debug views with my changes?

File: com.unity.render-pipelines.universal/Shaders/UnlitMetaPass.hlsl
Comment (nerites) [reply com.unity.render-pipelines.universal/Shaders/UnlitMetaPass.hlsl]: Fixed in 159a6ef

File: com.unity.render-pipelines.universal/Shaders/SimpleLitMetaPass.hlsl
Comment (nerites) [reply com.unity.render-pipelines.universal/Shaders/SimpleLitMetaPass.hlsl]: Fixed in 159a6ef

File: com.unity.render-pipelines.universal/ShaderLibrary/MetaInput.hlsl
Comment (nerites) [reply com.unity.render-pipelines.universal/ShaderLibrary/MetaInput.hlsl:28]: Fixed in 5b3637f

File: com.unity.render-pipelines.universal/ShaderLibrary/MetaInput.hlsl
Comment (nerites) [reply com.unity.render-pipelines.universal/ShaderLibrary/MetaInput.hlsl:21]: Fixed in 5b3637f

File: com.unity.render-pipelines.universal/ShaderLibrary/MetaInput.hlsl
Comment (nerites) [reply com.unity.render-pipelines.universal/ShaderLibrary/MetaInput.hlsl:47]: Fixed in 5b3637f

**Comment by nerites** [comment]
@phi-lira I ended up splitting the attributes/varyings declaration into a separate UniversalMetaPass.hlsl file, so that MetaInput.hlsl can be included in Shadergraph with conflicting with Shadergraph's Attributes/Varyings declarations. 

It's not great for backwards compatibility since it moves the Attributes and Varyings declarations into a new file, but the split seems in line with other *Input.hlsl/*Pass.hlsl content splits. What do you think?

6688d1f

File: com.unity.render-pipelines.high-definition/Runtime/Material/AxF/ShaderPass/AxFSharePass.hlsl
Comment (sebastienlagarde) [reply com.unity.render-pipelines.high-definition/Runtime/Material/AxF/ShaderPass/AxFSharePass.hlsl:20]: can you point me where UV3 is used in meta pass, I haven't noticed it? thanks

File: com.unity.render-pipelines.high-definition/Runtime/Material/Unlit/Unlit.hlsl
Comment (sebastienlagarde) [reply com.unity.render-pipelines.high-definition/Runtime/Material/Unlit/Unlit.hlsl:59]: yes we should revert this one

**Comment by sebastienlagarde** [comment]
> Do HDRP's debug modes also rely on the meta pass? Is it possible I've unintentionally modified HDRP's debug views with my changes?

nope, our debug mode are independent from Meta pass.

**Comment by nerites** [comment]
> 
> 
> > Do HDRP's debug modes also rely on the meta pass? Is it possible I've unintentionally modified HDRP's debug views with my changes?
> 
> nope, our debug mode are independent from Meta pass.
> 
> > It looks like different shaders handle emission differently, but I enabled emission maps wherever I could, and chose a default >checkerboard texture to use as the emission map. All emission colors are non-black, and all emission weights are 1.
> 
> Maybe it is due to exposure? In HDRP emissive is lighting and behave like a light, maybe you have different exposure setup? Also we ahve an option to have emissive as pre-expose or not.

Oh yeah, it looks like it was due to exposure- when I zoomed in close, I was able to see the emissive color I expected. I could see it even better when I disabled the exposure settings on my sky and fog volume.

> I don't get it, you mean that a lit material with emissive don't appear and that an unlit with a black emissive is visible?

I meant that both looked the same. Though with exposure disabled I see the emissive cube is brighter. The unlit cube with black emissive is still rendered in the debug mode though, is this intended?
image

File: com.unity.render-pipelines.high-definition/Runtime/Material/AxF/ShaderPass/AxFSharePass.hlsl
Comment (nerites) [reply com.unity.render-pipelines.high-definition/Runtime/Material/AxF/ShaderPass/AxFSharePass.hlsl:20]: Line 65 of ShaderPassLightTransport.hlsl

https://github.com/Unity-Technologies/Graphics/pull/4533/files#diff-8a05acf5191b37d00df7c8de3aa78907221288291e6062afc6c3bd71b7905876

Comment (nerites) [reply com.unity.render-pipelines.high-definition/Runtime/Material/AxF/ShaderPass/AxFSharePass.hlsl:20]: Hmm... I think I'm misunderstanding something about how the v2f struct is constructed. I assumed that each texcoord is a float2, but now I'm getting errors in HDRP shadergraphs because the texcoords seem to be float4s. If the texcoords can be float4s, then I would only need up to texcoord2.

**Comment by nerites** [comment]
Fixed metapass hookup in HDTarget for shadergraph!
Verified that emissive matches expectations from HDRP Debug view: 
image

File: com.unity.render-pipelines.high-definition/Runtime/Material/AxF/ShaderPass/AxFSharePass.hlsl
Comment (nerites) [reply com.unity.render-pipelines.high-definition/Runtime/Material/AxF/ShaderPass/AxFSharePass.hlsl:20]: ah, just assigning to .xy only fixes it. :) Can you check whether the usage of texcoords in ShaderPassLightTransport.hlsl is correct in latest?

File: com.unity.render-pipelines.high-definition/Runtime/Material/Unlit/Unlit.hlsl
Comment (nerites) [reply com.unity.render-pipelines.high-definition/Runtime/Material/Unlit/Unlit.hlsl]: Done in fe87b58

File: com.unity.render-pipelines.high-definition/Runtime/Material/AxF/ShaderPass/AxFSharePass.hlsl
Comment (sebastienlagarde) [reply com.unity.render-pipelines.high-definition/Runtime/Material/AxF/ShaderPass/AxFSharePass.hlsl:20]: ah right, so indeed we will keep the support of uv3 because it can be used in shader graph (i.E an artists could create a graph with fetching albedo with uv3 ). So indeed my bad, I have forge't about this :)

Comment (sebastienlagarde) [reply com.unity.render-pipelines.high-definition/Runtime/Material/AxF/ShaderPass/AxFSharePass.hlsl:20]: However you don't need to add the  ((SHADERPASS == SHADERPASS_LIGHT_TRANSPORT) as it is optional. it is IF the material use it then we will need to define it, so you should still remove this code

Comment (nerites) [reply com.unity.render-pipelines.high-definition/Runtime/Material/AxF/ShaderPass/AxFSharePass.hlsl:20]: Since the meta pass needs texcoord3, these lines in VertMesh.hlsl cause an error if uv3 does not also exist:

#if defined(VARYINGS_NEED_TEXCOORD3) || defined(VARYINGS_DS_NEED_TEXCOORD3)
    output.texCoord3 = input.uv3;
#endif

**Comment by nerites** [comment]
Holes don't get rendered in the sceneview debug modes in SRP, though they do get rendered for builtin. I'm still trying to get to the bottom of why this is happening, but renderdoc shows no holes texture is bound when rendering the meta passes in URP. I think this is to do with how it's not executing the SRP, but also not exactly executing the builtin render pipeline either... I'll do some more investigation tomorrow, but maybe this isn't worth fixing?

**Review by phocodes — APPROVED** [review]

**Review by erikabar — APPROVED** [review]
Baked Lightmap view is fixed on URP

File: com.unity.render-pipelines.universal/Shaders/Lit.shader
Comment (phi-lira) [reply com.unity.render-pipelines.universal/Shaders/Lit.shader:337]: I'm not sure what's the behaviour regarding shader_feature that's not references in any material, I guess it's stripped automatically. @erikabar maybe you can confirm this.

**Review by phi-lira — APPROVED** [review]

File: com.unity.render-pipelines.high-definition/Editor/Material/ShaderGraph/HDTarget.cs
Comment (nerites) [inline com.unity.render-pipelines.high-definition/Editor/Material/ShaderGraph/HDTarget.cs:608]: can remove texcoords since they will be added when corresponding uvs are used

File: com.unity.render-pipelines.high-definition/Runtime/Material/Lit/ShaderPass/LitSharePass.hlsl
Comment (nerites) [inline com.unity.render-pipelines.high-definition/Runtime/Material/Lit/ShaderPass/LitSharePass.hlsl:17]: convert all ((SHADERPASS == SHADERPASS_LIGHT_TRANSPORT) && defined(EDITOR_VISUALIZATION)) -> defined(EDITOR_VISUALIZATION))

File: com.unity.render-pipelines.high-definition/Editor/Material/ShaderGraph/HDTarget.cs
Comment (nerites) [reply com.unity.render-pipelines.high-definition/Editor/Material/ShaderGraph/HDTarget.cs:608]: If I remove these, the incorrect UVs are used to sample the debug texture, so I am leaving them in:
image

**Comment by nerites** [comment]
In testing shadergraphs, I've found that the hair, cotton, silk, and eye shadergraphs do not render in meta passes. However, Arnold, 3DS, and Autodesk do. Do you know of a difference between these shadergraphs that could be causing this @sebastienlagarde?

File: com.unity.render-pipelines.high-definition/Runtime/Material/Lit/ShaderPass/LitSharePass.hlsl
Comment (nerites) [reply com.unity.render-pipelines.high-definition/Runtime/Material/Lit/ShaderPass/LitSharePass.hlsl]: Done in 8611bcaa08

**Review by sebastienlagarde — APPROVED** [review]

**Comment by nerites** [comment]
Turns out I just forgot to enable GI on the MeshRenderers for those cubes, my bad! I did have to make a small change to get the hair SG working though. 
image

**Comment by nerites** [comment]
It looks like in URP the meta passes for ComplexLit, SimpleLit, BakedLit, and Unlit aren't getting picked up for some reason (they're rendered with the fallback GIDebug shader). I need some more information on shader keywords in order to get to the bottom of this. For now, I don't think it makes sense to block this PR on that, since other URP shaders are working (including shadergraphs), and the fallback shader should do just fine for non-instanced objects. 

If I get QA approval before I figure this out, I'll file a bug for the issue.

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/ShaderPass/ShaderPassLightTransport.hlsl
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/ShaderPass/ShaderPassLightTransport.hlsl:107]: oh sorry I missed this change in my first review. It is forbidden to access surfaceData in any of ShaderPassXXX.hlsl  it will not compile. Each Material can implement what they want.

What you need is to move the code sampling the Material here to each of the function GetLightTransportData() in Material it has been implemented for this.

What you need to do is to add SpecularColor in LightTransportData struct and move the code 
metaInput.SpecularColor = surfaceData.specularColor.rgb in each of the Material (each material have a different way to define spec color so you need to adapt for each material (you can check of the debug mode to output spec color convert the property for each material).

Then in this current files you will have only this:
UnityMetaInput metaInput;
    metaInput.Albedo = lightTransportData.diffuseColor.rgb;
    metaInput.Emission = lightTransportData.emissiveColor;
    metaInput.SpecularColor = lightTransportData.specularColor.rgb;
#endif
#ifdef EDITOR_VISUALIZATION
    metaInput.VizUV = input.texCoord1.xy;
    metaInput.LightCoord = float4(input.texCoord2.xy, input.texCoord3.xy);
#endif

**Comment by nerites** [comment]
Since dynamic GI has been fixed for HDRP in trunk, I was able to confirm that dynamic GI debug views also work, except for the emissive view. I still had to regenerate lighting a few times to get the other views working, and still saw the "failed to generate texture" message, which I will report to them.

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/ShaderPass/ShaderPassLightTransport.hlsl
Comment (nerites) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/ShaderPass/ShaderPassLightTransport.hlsl]: Ah actuaully, this SpecularColor was only used in validation debug views that are unsupported by SRPs (likely because their lighting models do not match builtin lighting models). So I just removed SpecularColor from MetaInput in fac9bab

**Comment by nerites** [comment]
Building URP_Terrain is consistently crashing. Attempting to build locally to repro.

File: com.unity.render-pipelines.universal/Shaders/Lit.shader
Comment (nerites) [reply com.unity.render-pipelines.universal/Shaders/Lit.shader:337]: It looks like you have to explicitly strip it. Kuba added it to his PR for stripping lighting-related keywords here: https://ono.unity3d.com/unity/unity/pull-request/127683/_/lighting/bugfix/1307896-dirlightmap-stripping

**Comment by sebastienlagarde** [comment]
Moving forward. Yamato is ok despite all the stuff broken, it come from our infrastructure and master

**Pull request merged**
